feat: EVM RPC Canister blog post (#2472)
dfinity/portal: blog/features/evm-rpc-canister-launch.mdx

<!-- prettier-ignore-start -->
:::info
Please ensure you [install](/docs/current/developer-docs/setup/install/) `dfx` before proceeding with the steps below.
:::
<!-- prettier-ignore-end -->

### Start a new project

```bash
dfx new evm_rpc_example
cd evm_rpc_example
```

### Add the EVM RPC Canister as a Dependency

Using `dfx deps` we can automatically install the EVM RPC canister wasm into our
local environment.

Edit the `dfx.json` file at the root of your project to add the EVM RPC canister
as a dependency under the "canisters" key:

```json
{
  "canisters": {
    "evm_rpc": {
      "type": "pull",
      "id": "7hfb6-caaaa-aaaar-qadga-cai"
    }
  },
  "defaults": {
    "build": {
      "args": "",
      "packtool": ""
    }
  },
  "output_env_file": ".env",
  "version": 1
}
```

Next, from your terminal, run the following command to fetch all dependencies.

```bash
dfx deps pull
```

Start the local replica

```bash
dfx start --background
```

Set the init args for the EVM RPC canister

```bash
dfx deps init evm_rpc --argument '(record { nodesInSubnet = 28 })'
```

Locally deploy the `evm_rpc` canister

```bash
dfx deps deploy evm_rpc
```

> Now, use `dfx` to get the current Ethereum gas price

Reviewer: ```suggestion
Now, use `dfx` to get the current Ethereum gas price:
```

```bash
dfx canister call evm_rpc request '(variant {Custom=record {url="https://ethereum.publicnode.com"}},"{\"jsonrpc\":\"2.0\",\"method\":\"eth_gasPrice\",\"params\":[],\"id\":1}",1000)' --wallet $(dfx identity get-wallet) --with-cycles 1000000000
```

The response:

```bash
(variant { Ok = "{\"id\":1,\"jsonrpc\":\"2.0\",\"result\":\"0x8ca25b481\"}" })
```

#### Raw RPC

You might be asking why the gas price is returned as a hex value. This endpoint
was reached using the generic `request` method expose by the EVM RPC canister.
This endpoint does not offer the same conveniences as the typed Candid RPC
endpoints, nor does it perform automatic consensus inference. This method is

> exposed mainly to give you access to any Ethereum RPC API method in case the

Reviewer: ```suggestion
exposed primarily to give you access to any Ethereum RPC API method in case the
```

typed Candid RPC endpoints are not sufficient.

<!-- prettier-ignore-start -->
:::info

> In the coming months, more tooling and improvements to the EVM RPC Canister are expected to land that will make it much easier to process raw Ethereum RPC data.

Reviewer: ```suggestion
In the coming months, more tooling and improvements to the EVM RPC canister are expected to be released that will make it much easier to process raw Ethereum RPC data.
```

:::
<!-- prettier-ignore-end -->

#### Candid RPC

Let's see what it looks like to call a typed Candid RPC method.

```bash
dfx canister call evm_rpc eth_getTransactionCount '(variant {EthMainnet = opt vec {variant {Cloudflare}}}, null, record {address = "0xdAC17F958D2ee523a2206206994597C13D831ec7"; block = variant {Latest}})' --with-cycles 1000000000 --wallet=$(dfx identity get-wallet)
```

Here, we call the `eth_getTransactionCount` method using Cloudflare as a
provider. This will return the number of transactions for the given contract
(i.e. address `0xdAC17F958D2ee523a2206206994597C13D831ec7`)

We get back the following, which tells us there is one transaction:

```bash
(variant { Consistent = variant { Ok = 1 : nat } })
```

## Conclusion

> The EVM RPC Canister is designed to help you build integrations with Ethereum

Reviewer: ```suggestion
The EVM RPC canister is designed to help you build integrations with Ethereum
```

smart contracts quickly, easily, and in a cost-effective manner. While it is
still early (and there are many improvements to come), this is a good chance to
get ahead and build slick ICP cross-chain dapps!

> If you have any feedback, or feel that the EVM RPC Canister is missing any

Reviewer: ```suggestion
If you have any feedback, or feel that the EVM RPC canister is missing any
```
